File: TP/palindromo.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

bool ehPalindromoRec(char palavra[], int i, int j) {
    if (i >= j) {
        return true;
    }
    if (palavra[i] != palavra[j]) {
        return false; 
    }
    return ehPalindromoRec(palavra, i + 1, j - 1); 
}

bool palavra(char palavra[]){
    return ehPalindromoRec(palavra, 0, strlen(palavra) - 1);
}

int main()
{
    char entrada[1000];
    int continuar = 1;  

    while (continuar == 1)
    {
        if (scanf(" %999[^\n]", entrada) != 1)
        {
            continuar = 0;
        }
        else
        {
            if (strcmp(entrada, "FIM") == 0)
            {
                continuar = 0; 
            }
            else
            {
                if (palavra(entrada))
                {
                    printf("SIM\n");
                }
                else
                {
                    printf("NAO\n");
                }
            }
        }
    }

    return 0;
}

File: TP/Palindromo.java
public class Palindromo {

    public static boolean ehPalindromo(String palavra) {
        int i = 0;
        int j = palavra.length() - 1;

        while (i < j) {
            if (palavra.charAt(i) != palavra.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

    public static void main(String[] args) {
        String palavra = MyIO.readLine();

        // Enquanto não for "FIM"
        while (!palavra.equals("FIM")) {
            if (ehPalindromo(palavra)) {
                MyIO.println("SIM");
            } else {
                MyIO.println("NAO");
            }

            palavra = MyIO.readLine(); 
        }
    }
}
